### ISSUE TITLE
Slow duplicated check build with global resources enabled

### ISSUE THREAD

**Issue by odormond** [concourse/concourse#7077]
## Summary

With global resources enabled, two pipelines with identical git resources will produce duplicated check builds every resource checking interval. One of which will run fine in a few seconds and the other will take ~1 minute to complete successfully but without any output. By "output", I mean that clicking on the check bar in the UI for the slow resource shows nothing.

## Steps to reproduce

Start a clean concourse install with a single linux worker. Started the ATC with `--enable-global-resource --resource-checking-interval 7m --container-placement-strategy limit-active-tasks --max-active-tasks-per-worker 1 --gc-failed-grace-period 1h` the rest is pretty much standard.

```
fly set-pipeline -p pipeline-1 -c simple-pipline-with-a-git-resource.yml
fly set-pipeline -p pipeline-2 -c simple-pipline-with-a-git-resource.yml
fly unpause-pipeline --all
watch -n 10 fly builds | grep check
```

## Expected results

Either a single check build for any pipeline defining the identical resource:
```
836  pipeline-2/resource.git/check      succeeded  2021-05-26@15:36:23+0200  2021-05-26@15:36:25+0200  2s        main  system
```

Or at least that none of the reported build check takes much longer than the other:
```
836  pipeline-2/resource.git/check      succeeded  2021-05-26@15:36:23+0200  2021-05-26@15:36:25+0200  2s        main  system
835  pipeline-1/resource.git/check      succeeded  2021-05-26@15:36:23+0200  2021-05-26@15:36:25+0200  2s        main  system
```

## Actual results

One of the two runs fast and the other stays there doing nothing for a minute or so:
```
836  pipeline-2/resource.git/check      succeeded  2021-05-26@15:36:23+0200  2021-05-26@15:36:25+0200  2s        main  system
835  pipeline-1/resource.git/check      succeeded  2021-05-26@15:36:23+0200  2021-05-26@15:37:23+0200  1m0s      main  system
```

## Additional context

Manually un-pausing the pipelines a few seconds apart resulted in a first check being effectively run on the worker and a second check build, which was not effectively executed, was reported as taking longer:
```
834  pipeline-2/resource.git/check      succeeded  2021-05-26@15:29:32+0200  2021-05-26@15:30:11+0200  39s       main  system
833  pipeline-1/resource.git/check      succeeded  2021-05-26@15:29:11+0200  2021-05-26@15:29:16+0200  5s        main  system
```

After waiting for the resource checking interval, two check builds are created at the same time:
```
836  pipeline-2/resource.git/check      succeeded  2021-05-26@15:36:23+0200  2021-05-26@15:36:25+0200  2s        main  system
835  pipeline-1/resource.git/check      succeeded  2021-05-26@15:36:23+0200  2021-05-26@15:37:23+0200  1m0s      main  system
```

Logs of the ATC (in debug) and the worker.
[concourse-web.log](https://github.com/concourse/concourse/files/6547531/concourse-web.log)
[concourse-worker.log](https://github.com/concourse/concourse/files/6547532/concourse-worker.log)

Disabling global resources fixes the problem but not for the very first check after the ATC is restarted. As can be seen in the logs below, check build ids 845 and 846 are initialized at 2021-05-26T14:12:09Z but 846 finishes at 2021-05-26T14:12:11Z while 845 is effectively started at 2021-05-26T14:13:01Z and finishes at 2021-05-26T14:13:05Z. Then they will each run in a few seconds, one minute apart. Note that they kept sharing the same container.
[concourse-web-no-g-r.log](https://github.com/concourse/concourse/files/6547670/concourse-web-no-g-r.log)

By the way, `fly containers` never update the build id it reports for a check container...
```
handle                                worker                       pipeline    job   build #  build id  type   name          attempt
b9e1736c-018c-4066-5b00-af3c76eafe2d  wrk-lin-1-0ea983261597099b9  pipeline-2  none  check    851       check  resource.git  n/a
```
While `fly builds` actually reports 860 after some times instead of 851:
```
860  pipeline-2/resource.git/check      succeeded  2021-05-26@17:02:22+0200  2021-05-26@17:02:24+0200  2s        main  system
```

## Triaging info

* Concourse version: 7.2.0
* Runtime: garden
* Browser (if applicable): Firefox but I don't think that matter as the `fly build` output exhibit the same delayed result for one of the pipeline.
* Did this used to work? Yes. We noticed these unresponsive check after updating from 6.7.2 straight to 7.2.0.

**Comment by evanchaoli** [comment]
That's because the second check was waiting for the first check to finish, then reuse result of the first check. For each global resource, there is a check log which force checks over a global resource to run in serial. Thus, the long time for the second check build was actually the time of waiting, it was not a real "slowness".

Before 7.x, say in your case, there are two pipelines with 1 global resource, in each check period, Concourse will create a 1 check in db; but since 7.x, Concourse will create 2 check builds in db, though only the first check will be really executed, 7.x way creates more check builds in db, executes more check builds (though most of those checks just do some db queries without involving check containers). From this perspective, 7.x has some performance downgrade. BTW, I have done #6854 which optimized checks a little bit, but not related to your issue.

I had a conversation with @vito. The 7.x way could be optimized, one possible solution is to not write check builds to db, or maybe there are other more solutions. I wanted to work on that, but never got a chance yet.

**Comment by odormond** [comment]
Having the second check wait for the result of the first one is perfectly fine, of course. But why wait a minute to notice the first one completed? It looks like their is actually a third element that runs every minute and which is the one that will notice the first check completed and notify the second.

One could argue that a minute delay is okayish, but for a dev sitting in front of his browser watching the yellow box to go green, it's a long wait.

Now, this is happening in a test setup with nothing else but these two pipelines running. In our production we have a much larger load and it's impossible to understand what's really happening. I'm guessing that the reported issue is part of the larger performance issue we have with concourse checks. To help you understand why we would really appreciate any gain, here are a few stats:
```
atc=# select type, min(execution), avg(execution), max(execution), count(*) from (select b.id, r.type, b.end_time - b.start_time as execution from builds b inner join resources r on b.resource_id = r.id where b.name = 'check' and b.status = 'succeeded') as checks group by type order by avg;
        type        |       min       |       avg       |       max       | count 
--------------------+-----------------+-----------------+-----------------+-------
 dynamic-metalink   | 00:00:01.719497 | 00:00:01.719497 | 00:00:01.719497 |     1
 smuggler           | 00:00:02.468727 | 00:00:02.468727 | 00:00:02.468727 |     1
 npm-cache          | 00:03:29.952744 | 00:03:29.952744 | 00:03:29.952744 |     1
 cron-resource      | 00:00:00.466626 | 00:03:35.884462 | 00:09:30.716868 |    85
 time               | 00:01:34.158661 | 00:05:33.767158 | 00:08:08.142086 |    29
 docker-image       | 00:07:40.349017 | 00:07:44.757024 | 00:08:06.662602 |    24
 counter-resource   | 00:05:04.289909 | 00:07:51.426234 | 00:08:28.470218 |    15
 git                | 00:00:01.585357 | 00:07:52.357648 | 00:08:46.448123 |   525
 semver             | 00:00:00.611615 | 00:07:53.543213 | 00:08:22.911386 |    81
 s3                 | 00:00:02.501578 | 00:07:56.518525 | 00:08:28.530873 |   404
 pull-request.image | 00:07:51.177019 | 00:08:00.938922 | 00:08:11.200068 |    10
 registry-image     | 00:07:41.16854  | 00:08:03.077999 | 00:09:24.504107 |   266
 s3-simple          | 00:08:15.12768  | 00:08:15.12768  | 00:08:15.12768  |     1
(13 rows)

atc=# select count(*) from pipelines;
 count 
-------
   251
(1 row)

atc=# select count(*) from resources;
 count 
-------
  4575
(1 row)

atc=# select count(*) from builds;
 count  
--------
 230336
(1 row)

atc=# select platform, count(*) from workers group by platform;
 platform | count 
----------+-------
 linux    |    27
 windows  |     6
 darwin   |    10
(3 rows)
```

As you can see, we basically have an 8 minutes delay whenever a pipeline is triggered because a check is run and it takes roughly 8 minutes on average to complete.

**Comment by evanchaoli** [comment]
@odormond I also hit this problem in my load test on 7.3.0, then I looked into the problem. It seems a Concourse bug. To workaround, you may try to set `CONCOURSE_MAX_CHECKS_PER_SECOND: -1`. Please let me know if this option solve the problem. I will submit a PR to fix the bug.

**Comment by evanchaoli** [comment]
BWT, I would recommend skip 7.2.0 and try 7.3.0 directly. 7.3.0 has an optimization to checks. Also, if all of your workers are Linux, consider to turn on the new feature `CONCOURSE_ENABLE_CACHE_STREAMED_VOLUMES: "true"`.

**Comment by evanchaoli** [comment]
> With check builds created for _every resource_ but spaced every resource check interval divided by the number of _resource config scopes_, which by definition is smaller or equal to the number of resources

Yes, that's exactly the bug I meant.

BTW, for cache streamed volumes, the only known issue is #7052. If you don't rely on `fly volumes` command, then you may consider to turn on and enjoy the feature. I saw Concourse team had enabled the feature on the Concourse itself's CI environment for long time. The feature will significantly reduce `git clone`s to your git system, and image pulls from dockerhub (as well as other image registries), which would benefit the entire ecosystem. Also, if a resource can be found locally, then `get` just does nothing, which will save containers and make builds faster. I think that is especially good for large deployments.

**Comment by odormond** [comment]
Hello!
Just a follow-up post to let you know that we disabled the throttling on Monday morning and haven't seen any issue since then. We even noticed that the overall build throughput increased:
![image](https://user-images.githubusercontent.com/6637092/120769300-45297380-c51d-11eb-8f9b-2c5cabc528d8.png)

We have also successfully tested 7.3.1 with cache streamed volumes enabled and will deploy it next week.
